File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */

/*
Given an integer array nums and an integer k
return the k most frequent elements

inputs: array nums and an integer k 
outputs: array nums
summarize: given an array of nums and an integer k we want to return the top k most frequent elements

*/

var topKFrequent = function(nums, k) {
    let output = []
    let freq = new Map() // O(n)
    let bucket = []
    // {1:3, 2:2, 3:1}
    // O(n)
    for (let num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1)
    }
    // buckets = [0,(6,7), (0,1,4,9,-4), (-3)]
    for (let [key, value] of freq) {
        if (bucket[value]) {
            bucket[value].push(key)
        } else {
            bucket[value] = [key]
        }
    }   
    // counts = [[3],[2],[1]]
    for (let i = bucket.length - 1; i >= 0; i--) {
       if (bucket[i]) output.push(...bucket[i])
        if (output.length === k) break
    }

    return output
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # inputs: integer array nums and an integer k
        # output: array of most frequent elements

        nums_length = len(nums)
        res = [[] for _ in range(nums_length + 1)]
        output = []

        counts = {}
        # 1:3, 2:2, 3:1
        # [0,[3],[2],[1],0,0]
        for num in nums:
            if num in counts:
                counts[num] += 1
            else:
                counts[num] = 1

        for a,v in counts.items():
            res[v].append(a)
        

        for i in range(len(res) - 1, -1, -1):
            if len(res[i]) > 0:
                for num in res[i]:
                    print(num, k)

                    output.append(num)
                    k -= 1
                    if k == 0:
                        return output
        return output